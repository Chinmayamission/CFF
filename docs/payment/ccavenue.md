# Initial setup
Before a CCAvenue account can be used with CFF, add a manual entry directly in the database with the required fields (this can be done from the Azure Cosmos DB portal).
```
{
    "merchant_id":"...",
    "SECRET_working_key":"...",
    "access_code":"...",
    "_cls":"chalicelib.models.CCAvenueConfig"
}
```

# Setup for a single form
Add the following in paymentMethods:
```
"paymentMethods": {
    "ccavenue": {
        "merchant_id": "...",
    }
}
```

## Sub account
To use a sub account, add the `sub_account_id` parameter as follows:

```
"paymentMethods": {
    "ccavenue": {
        "merchant_id": "...",
        "sub_account_id": "..."
    }
}
```