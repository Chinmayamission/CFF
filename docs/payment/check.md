To allow for payment by check, add the following key in paymentMethods:

```json
    "manual_approval": {
      "cff_show_when": "paymentMethod:check + registrationType:sponsorship",
      "payButtonText": "Pay by Check",
      "successMessage": "Hari Om,<br><br>Thank you for your registration. Make check payable to \"Chinmaya Mission San Jose\"<br>Please add Note: MSC 2020<br>10160 Clayton Rd, San Jose, CA 95127",
      "confirmationEmailInfo": {
        "template": {
          "html": "<div style='width: 100%;background-color: #eee; margin: 10px 0px;'> <div style='width: 80%;margin: auto; box-shadow: 1px 1px 4px grey;padding: 10px 30px;background: white;'> <img src='https://i.imgur.com/Cq2uKXU.png' style='width: 100%; max-width: 100px; margin: auto;'> <div style='width: 100%; max-width: 500px; text-align: center;'> <h1 style='margin: auto;'>Chinmaya Aradhana Camp 2020 Registration Confirmation</h1></div> <br> Hari OM, <br> <br> Thank you for signing up for the 2020 MSC. <Br> <table> <tr> <th>First Name</th> <td>{{value.contactName.first}}</td> </tr> <tr> <th>Last Name</th> <td>{{value.contactName.last}}</td> </tr> </table> <br> <h2>Payment Info</h2> <table> <tr> <th>Item Name</th> <th>Description</th> <th>Amount</th> </tr>{% for item in paymentInfo['items'] %} <tr> <td>{{item.name}}</td> <td>{{item.description}}</td> <td>{{item.amount | format_payment(paymentInfo.currency)}}</td> </tr>{% endfor %} </table> <br>Amount paid: {{amount_paid | format_payment(paymentInfo.currency)}} <br><br> You can view your response <a href='{{view_link}}'>at this link</a>. Make check payable to \"Chinmaya Mission San Jose\"<br>Please add Note: MSC 2020<br>10160 Clayton Rd, San Jose, CA 95127 <br><br> You can view your response <a href='{{view_link}}'>at this link</a>.<br><br> Thanks, <br> CMSJ Team <br><br> - </div> </div>"
        },
        "cc": [],
        "subject": "MSC payment pending - please mail your check",
        "toField": "email",
        "fromName": "MSC 2020 - Chinmaya Aradhana Camp 2020",
        "bcc": "parag@cmsj.org",
        "from": "MSC2020@cmsj.org"
      }
    }
```