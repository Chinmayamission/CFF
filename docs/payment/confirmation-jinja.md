## Template context

By default, all the variables for the given response are available through the `value` variable in the template context. For example, given a form response with the following value:

```json
{
    "name": "Ashwin"
}
```

Using the template `Hello, {{value.name}}` will resolve to `Hello, Ashwin`.

### View link

In the confirmation email template, you can use the `view_link` variable to get a response view link. This renders a form response which is disabled. Just do:

```html
Modify your response here: {{view_link}}.
```

The resulting view link will be equal to `{CFF_URL}/{formId}?responseId={responseId}&mode=view`.

!!! note
    The view link will only work if `formOptions.responseCanViewByLink` is set to true; this value is set to false by default for security reasons.

### Edit link

In the confirmation email template, you can use the `edit_link` variable to get a response edit link. This renders the form, with the response data pre-filled in. Just do:

```html
Modify your response here: {{edit_link}}.
```

The resulting view link will be equal to `{CFF_URL}/{formId}?responseId={responseId}&mode=edit`.

!!! note
    The edit link will only work if both `formOptions.responseCanViewByLink` and `formOptions.responseCanEditByLink` are set to true; both are set to false by default for security reasons.


## Custom filters

Jinja2 has a [filters](https://jinja.palletsprojects.com/en/2.11.x/templates/#filters) feature that allows you to apply certain functions to context variables within the template. In addition to the [built-in filters](https://jinja.palletsprojects.com/en/2.11.x/templates/#builtin-filters), CFF also supports the following custom filters:

### format_date
`format_date` formats a given date string (in "YYYY-MM-DD" format) to the "en" locale.

For example, the following template:

```html
{{"2000-10-10" | format_date}}
```

Gives a result of `Oct 10, 2000`.

### format_payment
`format_payment` formats a given amount in the specified currency.

For example, the following template:

```html
{{200 | format_payment("USD")}}
```

Gives a result of `$200`.

Note that this template is usually used with regard to `paymentInfo`, so a common use case might look something like this:

```html
{{paymentInfo.total | format_payment(paymentInfo.currency)}}
```