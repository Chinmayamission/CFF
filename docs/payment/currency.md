Specify a currency in `paymentInfo.currency`. The current currencies supported are USD and INR.

## Currency template

You can also specify `paymentInfo.currencyTemplate` to allow for a different currency based on the response value. The value of currencyTemplate must be a jinja2 template. For example:

```
{
    "currencyTemplate": "{% if value.nationality == 'India' %}INR{% else %}USD{% endif %}"
}
```

If the formData is `{"nationality": "India"}`, then the currency will become "INR"; else, "USD".

Note: We actually use (nunjucks)[https://github.com/mozilla/nunjucks] to parse the template on the frontend, so there may be specific compatibility differences between this and jinja2.