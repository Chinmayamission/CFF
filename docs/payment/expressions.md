There are many expressions understood by ExpresssionParser.

# Custom Functions

## cff_yeardiff
`cff_yeardiff` calculates the difference, in years, between the two dates are passed to it. Both dates have to be in YYYY-MM-DD format. A sample use case for this is calculating someone's age based on their date of birth.

For example, having the following for the expression:

```
cff_yeardiff('2019-09-01', dob)
```

When the form data is:

```
{"dob": '1999-01-12'}
``

Will make the final result evaluate to `20`.