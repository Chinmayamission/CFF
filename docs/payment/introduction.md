Forms can be integrated with payment methods. There are two JSON properties on formOptions that control payment:

- `formOptions.paymentInfo` - configures the actual payment calculation of the form. Based on the configuration specified in this field, you can configure how the final payment amounts are calculated based on form data, the payment currency, and more.

- `formOptions.paymentMethods` - configures the payment methods that the form is connected to. A form can be connected to a single payment method, multiple payment methods, or none.