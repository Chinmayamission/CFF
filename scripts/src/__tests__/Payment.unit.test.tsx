import React from "react";
import { shallow, mount, render } from "enzyme";
import Payment from "../form/confirmation/payment";
import { calculatePaymentInfo } from "../form/payment/PaymentCalcTable";

it("renders payment table on new response", () => {
  const wrapper = render(
    <Payment
      onPaymentStarted={e => e}
      paymentInfo={{
        currency: "USD",
        total: 12,
        items: [
          {
            name: "One",
            description: "One",
            amount: 12,
            quantity: 1,
            total: 12
          }
        ]
      }}
      paymentInfo_owed={{
        currency: "USD",
        total: 0
      }}
      paymentInfo_received={{
        currency: "USD",
        total: 0
      }}
      paymentMethods={[]}
      onPaymentComplete={e => e}
      onPaymentError={e => e}
      responseId={"responseId"}
      formId={"formId"}
      formData={{}}
    />
  );
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.text()).not.toContain("Amount already paid");
});

it("renders payment table with amount received", () => {
  const wrapper = render(
    <Payment
      onPaymentStarted={e => e}
      paymentInfo={{
        currency: "USD",
        total: 12,
        items: [
          {
            name: "One",
            description: "One",
            amount: 12,
            quantity: 1,
            total: 12
          }
        ]
      }}
      paymentInfo_owed={{
        currency: "USD",
        total: 7
      }}
      paymentInfo_received={{
        currency: "USD",
        total: 5
      }}
      paymentMethods={[]}
      onPaymentComplete={e => e}
      onPaymentError={e => e}
      responseId={"responseId"}
      formId={"formId"}
      formData={{}}
    />
  );
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.text()).toContain("Amount Already Paid: $5.00");
});

it("renders recurring payment table with times", () => {
  const wrapper = render(
    <Payment
      onPaymentStarted={e => e}
      paymentInfo={{
        currency: "USD",
        total: 12,
        items: [
          {
            name: "One",
            description: "One",
            amount: 12,
            quantity: 1,
            total: 12,
            recurrenceDuration: "1M"
          }
        ]
      }}
      paymentInfo_owed={{
        currency: "USD",
        total: 0
      }}
      paymentInfo_received={{
        currency: "USD",
        total: 0
      }}
      paymentMethods={[]}
      onPaymentComplete={e => e}
      onPaymentError={e => e}
      responseId={"responseId"}
      formId={"formId"}
      formData={{}}
    />
  );
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.text()).not.toContain("Amount already paid");
});

it("renders recurring payment table with end", () => {
  const wrapper = render(
    <Payment
      onPaymentStarted={e => e}
      paymentInfo={{
        currency: "USD",
        total: 120,
        items: [
          {
            name: "One",
            description: "One",
            amount: 12,
            quantity: 1,
            total: 120,
            recurrenceDuration: "1M",
            recurrenceTimes: "10"
          }
        ]
      }}
      paymentInfo_owed={{
        currency: "USD",
        total: 0
      }}
      paymentInfo_received={{
        currency: "USD",
        total: 0
      }}
      paymentMethods={[]}
      onPaymentComplete={e => e}
      onPaymentError={e => e}
      responseId={"responseId"}
      formId={"formId"}
      formData={{}}
    />
  );
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.text()).not.toContain("Amount already paid");
});

it("renders installment payment table", () => {
  const wrapper = render(
    <Payment
      onPaymentStarted={e => e}
      paymentInfo={{
        currency: "USD",
        total: 120,
        items: [
          {
            name: "One",
            description: "One",
            amount: 120,
            quantity: 1,
            total: 120
          },
          {
            name: "Installment",
            description: "Installment",
            amount: 12,
            quantity: 1,
            total: 120,
            recurrenceDuration: "1M",
            recurrenceTimes: "10",
            installment: true
          }
        ]
      }}
      paymentInfo_owed={{
        currency: "USD",
        total: 0
      }}
      paymentInfo_received={{
        currency: "USD",
        total: 0
      }}
      paymentMethods={[]}
      onPaymentComplete={e => e}
      onPaymentError={e => e}
      responseId={"responseId"}
      formId={"formId"}
      formData={{}}
    />
  );
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.text()).not.toContain("Amount already paid");
});

it("doesn't render payment table when paymentInfo.items is none", () => {
  const wrapper = render(
    <Payment
      onPaymentStarted={e => e}
      paymentInfo={{
        currency: "USD",
        total: 12,
        items: []
      }}
      paymentInfo_owed={{
        currency: "USD",
        total: 0
      }}
      paymentInfo_received={{
        currency: "USD",
        total: 0
      }}
      paymentMethods={[]}
      onPaymentComplete={e => e}
      onPaymentError={e => e}
      responseId={"responseId"}
      formId={"formId"}
      formData={{}}
    />
  );
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.text()).not.toContain("Description");
});

describe("calculatePaymentInfo()", () => {
  it("calculates simple paymentInfo", () => {
    let paymentInfo = {
      items: [
        {
          amount: "2",
          quantity: "2"
        }
      ]
    };
    let formData = {};
    expect(calculatePaymentInfo(paymentInfo, formData)).toMatchSnapshot();
  });
  it("calculates total items in paymentInfo last", () => {
    let paymentInfo = {
      items: [
        {
          amount: "2",
          quantity: "2"
        },
        {
          amount: "-0.1 * $total",
          quantity: "1"
        }
      ]
    };
    let formData = {};
    expect(calculatePaymentInfo(paymentInfo, formData)).toMatchSnapshot();
  });
  it("doesn't include installments in total calculation", () => {
    let paymentInfo = {
      items: [
        {
          amount: "2",
          quantity: "2"
        },
        {
          amount: "0.5 * $total",
          quantity: "1",
          recurrenceDuration: "1M",
          recurrenceTimes: 2,
          installment: true
        }
      ]
    };
    let formData = {};
    expect(calculatePaymentInfo(paymentInfo, formData)).toMatchSnapshot();
  });
  it("calculates installments from multiple totals", () => {
    let paymentInfo = {
      items: [
        {
          amount: "2",
          quantity: "2"
        },
        {
          amount: "-0.25 * $total",
          quantity: "1"
        },
        {
          amount: "-0.25 * $total",
          quantity: "1"
        },
        {
          amount: "0.5 * $total",
          quantity: "1",
          recurrenceDuration: "1M",
          recurrenceTimes: 2,
          installment: true
        }
      ]
    };
    let formData = {};
    expect(calculatePaymentInfo(paymentInfo, formData)).toMatchSnapshot();
  });
});
