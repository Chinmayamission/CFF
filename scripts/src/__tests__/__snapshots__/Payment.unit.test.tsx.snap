// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculatePaymentInfo() calculates installments from multiple totals 1`] = `
Object {
  "items": Array [
    Object {
      "amount": 2,
      "quantity": 2,
      "total": 4,
    },
    Object {
      "amount": -1,
      "quantity": 1,
      "total": -1,
    },
    Object {
      "amount": -1,
      "quantity": 1,
      "total": -1,
    },
    Object {
      "amount": 1,
      "installment": true,
      "quantity": 1,
      "recurrenceDuration": "1M",
      "recurrenceTimes": 2,
      "total": 2,
    },
  ],
  "total": 2,
}
`;

exports[`calculatePaymentInfo() calculates simple paymentInfo 1`] = `
Object {
  "items": Array [
    Object {
      "amount": 2,
      "quantity": 2,
      "total": 4,
    },
  ],
  "total": 4,
}
`;

exports[`calculatePaymentInfo() calculates total items in paymentInfo last 1`] = `
Object {
  "items": Array [
    Object {
      "amount": 2,
      "quantity": 2,
      "total": 4,
    },
    Object {
      "amount": -0.4,
      "quantity": 1,
      "total": -0.4,
    },
  ],
  "total": 3.6,
}
`;

exports[`calculatePaymentInfo() doesn't include installments in total calculation 1`] = `
Object {
  "items": Array [
    Object {
      "amount": 2,
      "quantity": 2,
      "total": 4,
    },
    Object {
      "amount": 2,
      "installment": true,
      "quantity": 1,
      "recurrenceDuration": "1M",
      "recurrenceTimes": 2,
      "total": 4,
    },
  ],
  "total": 4,
}
`;

exports[`doesn't render payment table when paymentInfo.items is none 1`] = `
<div>
  <h1
    class="text-center my-2"
  >
    Payment
  </h1>
  <div />
  <div>
    <div
      class="alert alert-info"
    >
      We have already received your payment. No additional payment necessary -- you will receive a confirmation email shortly about your update.
    </div>
  </div>
</div>
`;

exports[`renders installment payment table 1`] = `
<div>
  <h1
    class="text-center my-2"
  >
    Payment
  </h1>
  <div>
    <div>
      <div
        class="mb-2"
      >
        <div
          class="ReactTable my-4"
        >
          <div
            class="rt-table"
            role="grid"
          >
            <div
              class="rt-thead -header"
              style="min-width:500px"
            >
              <div
                class="rt-tr"
                role="row"
              >
                <div
                  class="rt-th rt-resizable-header -cursor-pointer"
                  role="columnheader"
                  style="flex:100 0 auto;width:100px"
                  tabindex="-1"
                >
                  <div
                    class="rt-resizable-header-content"
                  >
                    Name
                  </div>
                  <div
                    class="rt-resizer"
                  />
                </div>
                <div
                  class="rt-th rt-resizable-header -cursor-pointer"
                  role="columnheader"
                  style="flex:100 0 auto;width:100px"
                  tabindex="-1"
                >
                  <div
                    class="rt-resizable-header-content"
                  >
                    Description
                  </div>
                  <div
                    class="rt-resizer"
                  />
                </div>
                <div
                  class="rt-th rt-resizable-header -cursor-pointer"
                  role="columnheader"
                  style="flex:100 0 auto;width:100px"
                  tabindex="-1"
                >
                  <div
                    class="rt-resizable-header-content"
                  >
                    Amount
                  </div>
                  <div
                    class="rt-resizer"
                  />
                </div>
                <div
                  class="rt-th rt-resizable-header -cursor-pointer"
                  role="columnheader"
                  style="flex:100 0 auto;width:100px"
                  tabindex="-1"
                >
                  <div
                    class="rt-resizable-header-content"
                  >
                    Quantity
                  </div>
                  <div
                    class="rt-resizer"
                  />
                </div>
                <div
                  class="rt-th rt-resizable-header -cursor-pointer"
                  role="columnheader"
                  style="flex:100 0 auto;width:100px"
                  tabindex="-1"
                >
                  <div
                    class="rt-resizable-header-content"
                  >
                    Total
                  </div>
                  <div
                    class="rt-resizer"
                  />
                </div>
              </div>
            </div>
            <div
              class="rt-tbody"
              style="min-width:500px"
            >
              <div
                class="rt-tr-group"
                role="rowgroup"
              >
                <div
                  class="rt-tr -odd"
                  role="row"
                >
                  <div
                    class="rt-td"
                    role="gridcell"
                    style="flex:100 0 auto;width:100px"
                  >
                    One
                  </div>
                  <div
                    class="rt-td"
                    role="gridcell"
                    style="flex:100 0 auto;width:100px"
                  >
                    One
                  </div>
                  <div
                    class="rt-td"
                    role="gridcell"
                    style="text-align:right;flex:100 0 auto;width:100px"
                  >
                    $120.00
                  </div>
                  <div
                    class="rt-td"
                    role="gridcell"
                    style="text-align:right;flex:100 0 auto;width:100px"
                  >
                    1
                  </div>
                  <div
                    class="rt-td"
                    role="gridcell"
                    style="text-align:right;flex:100 0 auto;width:100px"
                  >
                    $120.00
                  </div>
                </div>
              </div>
              <div
                class="rt-tr-group"
                role="rowgroup"
              >
                <div
                  class="rt-tr -even"
                  role="row"
                >
                  <div
                    class="rt-td"
                    role="gridcell"
                    style="flex:100 0 auto;width:100px"
                  >
                    Installment
                  </div>
                  <div
                    class="rt-td"
                    role="gridcell"
                    style="flex:100 0 auto;width:100px"
                  >
                    Installment
                  </div>
                  <div
                    class="rt-td"
                    role="gridcell"
                    style="text-align:right;flex:100 0 auto;width:100px"
                  >
                    $12.00
                  </div>
                  <div
                    class="rt-td"
                    role="gridcell"
                    style="text-align:right;flex:100 0 auto;width:100px"
                  >
                     every month for 10 times
                  </div>
                  <div
                    class="rt-td"
                    role="gridcell"
                    style="text-align:right;flex:100 0 auto;width:100px"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="-loading"
          >
            <div
              class="-loading-inner"
            >
              Loading...
            </div>
          </div>
        </div>
        Total Amount: $120.00
      </div>
    </div>
  </div>
  <div>
    <div
      class="alert alert-info"
    >
      We have already received your payment. No additional payment necessary -- you will receive a confirmation email shortly about your update.
    </div>
  </div>
</div>
`;

exports[`renders payment table on new response 1`] = `
<div>
  <h1
    class="text-center my-2"
  >
    Payment
  </h1>
  <div>
    <div>
      <div
        class="mb-2"
      >
        <div
          class="ReactTable my-4"
        >
          <div
            class="rt-table"
            role="grid"
          >
            <div
              class="rt-thead -header"
              style="min-width:500px"
            >
              <div
                class="rt-tr"
                role="row"
              >
                <div
                  class="rt-th rt-resizable-header -cursor-pointer"
                  role="columnheader"
                  style="flex:100 0 auto;width:100px"
                  tabindex="-1"
                >
                  <div
                    class="rt-resizable-header-content"
                  >
                    Name
                  </div>
                  <div
                    class="rt-resizer"
                  />
                </div>
                <div
                  class="rt-th rt-resizable-header -cursor-pointer"
                  role="columnheader"
                  style="flex:100 0 auto;width:100px"
                  tabindex="-1"
                >
                  <div
                    class="rt-resizable-header-content"
                  >
                    Description
                  </div>
                  <div
                    class="rt-resizer"
                  />
                </div>
                <div
                  class="rt-th rt-resizable-header -cursor-pointer"
                  role="columnheader"
                  style="flex:100 0 auto;width:100px"
                  tabindex="-1"
                >
                  <div
                    class="rt-resizable-header-content"
                  >
                    Amount
                  </div>
                  <div
                    class="rt-resizer"
                  />
                </div>
                <div
                  class="rt-th rt-resizable-header -cursor-pointer"
                  role="columnheader"
                  style="flex:100 0 auto;width:100px"
                  tabindex="-1"
                >
                  <div
                    class="rt-resizable-header-content"
                  >
                    Quantity
                  </div>
                  <div
                    class="rt-resizer"
                  />
                </div>
                <div
                  class="rt-th rt-resizable-header -cursor-pointer"
                  role="columnheader"
                  style="flex:100 0 auto;width:100px"
                  tabindex="-1"
                >
                  <div
                    class="rt-resizable-header-content"
                  >
                    Total
                  </div>
                  <div
                    class="rt-resizer"
                  />
                </div>
              </div>
            </div>
            <div
              class="rt-tbody"
              style="min-width:500px"
            >
              <div
                class="rt-tr-group"
                role="rowgroup"
              >
                <div
                  class="rt-tr -odd"
                  role="row"
                >
                  <div
                    class="rt-td"
                    role="gridcell"
                    style="flex:100 0 auto;width:100px"
                  >
                    One
                  </div>
                  <div
                    class="rt-td"
                    role="gridcell"
                    style="flex:100 0 auto;width:100px"
                  >
                    One
                  </div>
                  <div
                    class="rt-td"
                    role="gridcell"
                    style="text-align:right;flex:100 0 auto;width:100px"
                  >
                    $12.00
                  </div>
                  <div
                    class="rt-td"
                    role="gridcell"
                    style="text-align:right;flex:100 0 auto;width:100px"
                  >
                    1
                  </div>
                  <div
                    class="rt-td"
                    role="gridcell"
                    style="text-align:right;flex:100 0 auto;width:100px"
                  >
                    $12.00
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="-loading"
          >
            <div
              class="-loading-inner"
            >
              Loading...
            </div>
          </div>
        </div>
        Total Amount: $12.00
      </div>
    </div>
  </div>
  <div>
    <div
      class="alert alert-info"
    >
      We have already received your payment. No additional payment necessary -- you will receive a confirmation email shortly about your update.
    </div>
  </div>
</div>
`;

exports[`renders payment table with amount received 1`] = `
<div>
  <h1
    class="text-center my-2"
  >
    Payment
  </h1>
  <div>
    <div>
      <div
        class="mb-2"
      >
        <div
          class="ReactTable my-4"
        >
          <div
            class="rt-table"
            role="grid"
          >
            <div
              class="rt-thead -header"
              style="min-width:500px"
            >
              <div
                class="rt-tr"
                role="row"
              >
                <div
                  class="rt-th rt-resizable-header -cursor-pointer"
                  role="columnheader"
                  style="flex:100 0 auto;width:100px"
                  tabindex="-1"
                >
                  <div
                    class="rt-resizable-header-content"
                  >
                    Name
                  </div>
                  <div
                    class="rt-resizer"
                  />
                </div>
                <div
                  class="rt-th rt-resizable-header -cursor-pointer"
                  role="columnheader"
                  style="flex:100 0 auto;width:100px"
                  tabindex="-1"
                >
                  <div
                    class="rt-resizable-header-content"
                  >
                    Description
                  </div>
                  <div
                    class="rt-resizer"
                  />
                </div>
                <div
                  class="rt-th rt-resizable-header -cursor-pointer"
                  role="columnheader"
                  style="flex:100 0 auto;width:100px"
                  tabindex="-1"
                >
                  <div
                    class="rt-resizable-header-content"
                  >
                    Amount
                  </div>
                  <div
                    class="rt-resizer"
                  />
                </div>
                <div
                  class="rt-th rt-resizable-header -cursor-pointer"
                  role="columnheader"
                  style="flex:100 0 auto;width:100px"
                  tabindex="-1"
                >
                  <div
                    class="rt-resizable-header-content"
                  >
                    Quantity
                  </div>
                  <div
                    class="rt-resizer"
                  />
                </div>
                <div
                  class="rt-th rt-resizable-header -cursor-pointer"
                  role="columnheader"
                  style="flex:100 0 auto;width:100px"
                  tabindex="-1"
                >
                  <div
                    class="rt-resizable-header-content"
                  >
                    Total
                  </div>
                  <div
                    class="rt-resizer"
                  />
                </div>
              </div>
            </div>
            <div
              class="rt-tbody"
              style="min-width:500px"
            >
              <div
                class="rt-tr-group"
                role="rowgroup"
              >
                <div
                  class="rt-tr -odd"
                  role="row"
                >
                  <div
                    class="rt-td"
                    role="gridcell"
                    style="flex:100 0 auto;width:100px"
                  >
                    One
                  </div>
                  <div
                    class="rt-td"
                    role="gridcell"
                    style="flex:100 0 auto;width:100px"
                  >
                    One
                  </div>
                  <div
                    class="rt-td"
                    role="gridcell"
                    style="text-align:right;flex:100 0 auto;width:100px"
                  >
                    $12.00
                  </div>
                  <div
                    class="rt-td"
                    role="gridcell"
                    style="text-align:right;flex:100 0 auto;width:100px"
                  >
                    1
                  </div>
                  <div
                    class="rt-td"
                    role="gridcell"
                    style="text-align:right;flex:100 0 auto;width:100px"
                  >
                    $12.00
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="-loading"
          >
            <div
              class="-loading-inner"
            >
              Loading...
            </div>
          </div>
        </div>
        Total Amount: $12.00
      </div>
    </div>
    <div>
      <div>
        <div>
          Amount Already Paid: $5.00
        </div>
      </div>
      <div>
        <b>
          Amount Owed: $7.00 
        </b>
      </div>
    </div>
  </div>
  <div>
    <div
      class="container-fluid row"
      style="text-align:center"
    >
      <div
        class="col-12"
      >
        Please select a payment method to complete the form. You will receive a confirmation email after the payment is complete.
      </div>
    </div>
  </div>
</div>
`;

exports[`renders recurring payment table with end 1`] = `
<div>
  <h1
    class="text-center my-2"
  >
    Payment
  </h1>
  <div>
    <div>
      <div
        class="mb-2"
      >
        <div
          class="ReactTable my-4"
        >
          <div
            class="rt-table"
            role="grid"
          >
            <div
              class="rt-thead -header"
              style="min-width:500px"
            >
              <div
                class="rt-tr"
                role="row"
              >
                <div
                  class="rt-th rt-resizable-header -cursor-pointer"
                  role="columnheader"
                  style="flex:100 0 auto;width:100px"
                  tabindex="-1"
                >
                  <div
                    class="rt-resizable-header-content"
                  >
                    Name
                  </div>
                  <div
                    class="rt-resizer"
                  />
                </div>
                <div
                  class="rt-th rt-resizable-header -cursor-pointer"
                  role="columnheader"
                  style="flex:100 0 auto;width:100px"
                  tabindex="-1"
                >
                  <div
                    class="rt-resizable-header-content"
                  >
                    Description
                  </div>
                  <div
                    class="rt-resizer"
                  />
                </div>
                <div
                  class="rt-th rt-resizable-header -cursor-pointer"
                  role="columnheader"
                  style="flex:100 0 auto;width:100px"
                  tabindex="-1"
                >
                  <div
                    class="rt-resizable-header-content"
                  >
                    Amount
                  </div>
                  <div
                    class="rt-resizer"
                  />
                </div>
                <div
                  class="rt-th rt-resizable-header -cursor-pointer"
                  role="columnheader"
                  style="flex:100 0 auto;width:100px"
                  tabindex="-1"
                >
                  <div
                    class="rt-resizable-header-content"
                  >
                    Quantity
                  </div>
                  <div
                    class="rt-resizer"
                  />
                </div>
                <div
                  class="rt-th rt-resizable-header -cursor-pointer"
                  role="columnheader"
                  style="flex:100 0 auto;width:100px"
                  tabindex="-1"
                >
                  <div
                    class="rt-resizable-header-content"
                  >
                    Total
                  </div>
                  <div
                    class="rt-resizer"
                  />
                </div>
              </div>
            </div>
            <div
              class="rt-tbody"
              style="min-width:500px"
            >
              <div
                class="rt-tr-group"
                role="rowgroup"
              >
                <div
                  class="rt-tr -odd"
                  role="row"
                >
                  <div
                    class="rt-td"
                    role="gridcell"
                    style="flex:100 0 auto;width:100px"
                  >
                    One
                  </div>
                  <div
                    class="rt-td"
                    role="gridcell"
                    style="flex:100 0 auto;width:100px"
                  >
                    One
                  </div>
                  <div
                    class="rt-td"
                    role="gridcell"
                    style="text-align:right;flex:100 0 auto;width:100px"
                  >
                    $12.00
                  </div>
                  <div
                    class="rt-td"
                    role="gridcell"
                    style="text-align:right;flex:100 0 auto;width:100px"
                  >
                     every month for 10 times
                  </div>
                  <div
                    class="rt-td"
                    role="gridcell"
                    style="text-align:right;flex:100 0 auto;width:100px"
                  >
                    $120.00
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="-loading"
          >
            <div
              class="-loading-inner"
            >
              Loading...
            </div>
          </div>
        </div>
        Total Amount: $120.00
      </div>
    </div>
  </div>
  <div>
    <div
      class="alert alert-info"
    >
      We have already received your payment. No additional payment necessary -- you will receive a confirmation email shortly about your update.
    </div>
  </div>
</div>
`;

exports[`renders recurring payment table with times 1`] = `
<div>
  <h1
    class="text-center my-2"
  >
    Payment
  </h1>
  <div>
    <div>
      <div
        class="mb-2"
      >
        <div
          class="ReactTable my-4"
        >
          <div
            class="rt-table"
            role="grid"
          >
            <div
              class="rt-thead -header"
              style="min-width:500px"
            >
              <div
                class="rt-tr"
                role="row"
              >
                <div
                  class="rt-th rt-resizable-header -cursor-pointer"
                  role="columnheader"
                  style="flex:100 0 auto;width:100px"
                  tabindex="-1"
                >
                  <div
                    class="rt-resizable-header-content"
                  >
                    Name
                  </div>
                  <div
                    class="rt-resizer"
                  />
                </div>
                <div
                  class="rt-th rt-resizable-header -cursor-pointer"
                  role="columnheader"
                  style="flex:100 0 auto;width:100px"
                  tabindex="-1"
                >
                  <div
                    class="rt-resizable-header-content"
                  >
                    Description
                  </div>
                  <div
                    class="rt-resizer"
                  />
                </div>
                <div
                  class="rt-th rt-resizable-header -cursor-pointer"
                  role="columnheader"
                  style="flex:100 0 auto;width:100px"
                  tabindex="-1"
                >
                  <div
                    class="rt-resizable-header-content"
                  >
                    Amount
                  </div>
                  <div
                    class="rt-resizer"
                  />
                </div>
                <div
                  class="rt-th rt-resizable-header -cursor-pointer"
                  role="columnheader"
                  style="flex:100 0 auto;width:100px"
                  tabindex="-1"
                >
                  <div
                    class="rt-resizable-header-content"
                  >
                    Quantity
                  </div>
                  <div
                    class="rt-resizer"
                  />
                </div>
                <div
                  class="rt-th rt-resizable-header -cursor-pointer"
                  role="columnheader"
                  style="flex:100 0 auto;width:100px"
                  tabindex="-1"
                >
                  <div
                    class="rt-resizable-header-content"
                  >
                    Total
                  </div>
                  <div
                    class="rt-resizer"
                  />
                </div>
              </div>
            </div>
            <div
              class="rt-tbody"
              style="min-width:500px"
            >
              <div
                class="rt-tr-group"
                role="rowgroup"
              >
                <div
                  class="rt-tr -odd"
                  role="row"
                >
                  <div
                    class="rt-td"
                    role="gridcell"
                    style="flex:100 0 auto;width:100px"
                  >
                    One
                  </div>
                  <div
                    class="rt-td"
                    role="gridcell"
                    style="flex:100 0 auto;width:100px"
                  >
                    One
                  </div>
                  <div
                    class="rt-td"
                    role="gridcell"
                    style="text-align:right;flex:100 0 auto;width:100px"
                  >
                    $12.00
                  </div>
                  <div
                    class="rt-td"
                    role="gridcell"
                    style="text-align:right;flex:100 0 auto;width:100px"
                  >
                     every month
                  </div>
                  <div
                    class="rt-td"
                    role="gridcell"
                    style="text-align:right;flex:100 0 auto;width:100px"
                  >
                    $12.00
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="-loading"
          >
            <div
              class="-loading-inner"
            >
              Loading...
            </div>
          </div>
        </div>
        Total Amount: $12.00
      </div>
    </div>
  </div>
  <div>
    <div
      class="alert alert-info"
    >
      We have already received your payment. No additional payment necessary -- you will receive a confirmation email shortly about your update.
    </div>
  </div>
</div>
`;
